File: 2020/01/Part1.py
from typing import List

def read_input(name="input.txt") -> List[int]:
    l = []
    with open(name) as f:
        for line in f:
            if line.strip() != "":
                l.append(int(line.strip()))
    return l

def brute_force(l: List[int]) -> int:
    """
    Runs in quadratic time
    """

    for index1 in range(len(l)):
        for index2 in range(index1+1, len(l)):
            if l[index1] + l[index2] == 2020:
                return l[index1] * l[index2]
    return 0

def binary_search(l: List[int], target: int) -> bool:

    start = 0
    end = len(l) - 1
    while start <= end:
        mid = (start + end) // 2
        if l[mid] == target:
            return True
        elif l[mid] < target:
            start = mid + 1
        else:
            end = mid - 1

    return False

def sort_search(l: List[int]) -> int:
    """
    Idea is to sort a copy of the list. Then we can iterate through the
    first list in order and binary search for 2020- that element in the 
    sorted list. That should bring the asymptotic running time down to
    linearithmic
    """

    s = sorted(l) # sort into ascending order

    for item in l:
        opposite = 2020 - item
        if binary_search(s, opposite):
            return opposite * item
    return 0

def main():
    l = read_input()
    print(sort_search(l))

def test():

    l = read_input("input_test.txt")

    assert brute_force(l) == 514579, "Brute force failed on test input"
    assert sort_search(l) == 514579, "Sort search failed on test input"

    print("All tests passed")

if __name__ == "__main__":
    from sys import argv
    if len(argv) >= 2 and argv[1] == "test":
        test()
    else:
        main()

File: 2020/15/Part1.java
public class Part1 {

    private static void test() {

        Shared.myAssert(Shared.solve("0,3,6", 2020) == 436L, "Failed test input 0");
        Shared.myAssert(Shared.solve("1,3,2", 2020) == 1L, "Failed test input 1");
        Shared.myAssert(Shared.solve("2,1,3", 2020) == 10L, "Failed test input 2");
        Shared.myAssert(Shared.solve("1,2,3", 2020) == 27L, "Failed test input 3");
        Shared.myAssert(Shared.solve("2,3,1", 2020) == 78L, "Failed test input 4");
        Shared.myAssert(Shared.solve("3,2,1", 2020) == 438L, "Failed test input 5");
        Shared.myAssert(Shared.solve("3,1,2", 2020) == 1836L, "Failed test input 6");

        System.out.println("All tests passed");
    }

    public static void main(String[] args) {
        if (args.length >= 1 && args[0].equals("test")) {
            test();
        } else {
            System.out.println(Shared.solve("10,16,6,0,1,17", 2020));
        }
    }
}